compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -436,6 +436,39 @@ class Objects(using Context @constructorOnly):
       case _ =>
         throw new RuntimeException("Incorrect local environment for initializing " + x.show)
 
+    /**
+     * Resolve the environment by searching for a given symbol.
+     *
+     * Searches for the environment that owns `target`, starting from `env` as the innermost.
+     *
+     * Due to widening, the corresponding environment might not exist. As a result reading the local
+     * variable will return `Cold` and it's forbidden to write to the local variable.
+     *
+     * @param target  The symbol to search for.
+     * @param thisV The value for `this` of the enclosing class where the local variable is referenced.
+     * @param env   The local environment where the local variable is referenced.
+     *
+     * @return the environment that owns the `target` and value for `this` owned by the given method.
+     */
+    def resolveEnvByValue(target: Symbol, thisV: ThisValue, env: Data)(using Context): Option[(ThisValue, Data)] = log("Resolving env by value for " + target.show + ", this = " + thisV.show + ", env = " + env.show, printer) {
+      env match
+      case localEnv: LocalEnv =>
+        if localEnv.getVal(target).isDefined then Some(thisV -> localEnv)
+        else if localEnv.getVar(target).isDefined then Some(thisV -> localEnv)
+        else resolveEnvByValue(target, thisV, localEnv.outer)
+      case NoEnv =>
+        thisV match
+        case ref: OfClass =>
+          ref.outer match
+          case outer : ThisValue =>
+            resolveEnvByValue(target, outer, ref.env)
+          case _ =>
+            // TODO: properly handle the case where ref.outer is ValueSet
+            None
+        case _ =>
+          None
+    }
```

Comment:
I was thinking whether we can simplify `resolveEnvByValue` to just return the value for the local variable. We'll probably need two methods, one for mutable, one for immutable. But the logic would be simpler and it'll be easier to handle the TODOs.

That leaves us with the case for lazy local variables, for that one, we can use `resolveEnvByOwner`.

If it makes sense, it can be experimented in another PR.